algosdk/future/atomic_transaction_composer.py

```diff
@@ -0,0 +1,373 @@
+                method has no return value (void), then the method results array
+                will contain None for that method's return value.
+        """
+        if self.status > AtomicTransactionComposerStatus.SUBMITTED:
+            raise error.AtomicTransactionComposerError(
+                "AtomicTransactionComposerStatus must be submitted or lower to execute a group"
+            )
+
+        client.submit(client)
+        client.wait_for_confirmation(client, self.tx_ids[0])
+
+        resp = client.pending_transaction_info(self.tx_ids[0])
+        resp_dict = json.loads(resp)
+
+        confirmed_round = resp_dict["confirmed-round"]
+        tx_ids = self.tx_ids
+        results = resp_dict["logs"] if "logs" in resp_dict else None
+
+        method_results = []
+        for i, result in enumerate(results):
+            if self.method_list[i].Returns.type == "void":
+                method_results.append(None)
+            else:
+                return_value = self.method_list[i].Returns.type.decode(result)
+                method_results.append(return_value)
+
+        self.status = AtomicTransactionComposerStatus.COMMITTED
+
+        return AtomicTransactionResponse(
+            confirmed_round=confirmed_round,
+            tx_ids=tx_ids,
+            results=method_results,
+        )
+
+
+class TransactionSigner:
+    """
+    Represents a function which can sign transactions from an atomic transaction group.
+
+    Args:
+        private_key (str): private key of signing account
+        msig (MultiSig, optional): multisig account information
+    """
+
+    def __init__(self, private_key, msig=None) -> None:
+        self.private_key = private_key
+        self.msig = msig
+
+    def sign(self, txn_group, indexes):
+        signed_txn = []
+        for i in indexes:
+            txn = txn_group[i]
+            assert isinstance(txn, transaction.Transaction)
+            signed_txn.append(self.sign_txn(txn))
+        return signed_txn
+
+    def sign_txn(self, txn):
+        if self.msig:
+            for sk in self.private_key:
+                txn.sign(sk)
+            return txn
+        return txn.sign(self.private_key)
+
+
+class TransactionWithSigner:
+    def __init__(self, txn, signer) -> None:
+        self.txn = txn
+        self.signer = signer
+
+
+class AtomicTransactionResponse:
+    def __init__(self, confirmed_round, tx_ids=None, results=None) -> None:
+        self.confirmed_round = confirmed_round
+        self.tx_ids = tx_ids
+        self.method_results = results
+
+
+def make_account_transaction_signer(private_key):
+    """
+    Creates a TransactionSigner for a basic account or LogicSigAccount
+    """
+    return TransactionSigner(private_key)
+
+
+def make_multisig_account_transaction_signer(msig, sks):
+    """
+    Creates a TransactionSigner for a MultiSig account
+    """
+    return TransactionSigner(private_key=sks, msig=msig)
```

Comment:
nit: I guess you can put them into `TransactionSigner` as a class method? See: https://stackoverflow.com/questions/682504/what-is-a-clean-pythonic-way-to-have-multiple-constructors-in-python